Storage: Support iterating over NMaps with partial keys

tadeohepperle opened the pull request:
fixes #1044 

Changes to the Codegen to support iterating over partial keys.

Instead of 
```rs
runtime::storage().pallet().entry(key1, key2, key3); // fetchable and iterable (redirects to entry_root())
runtime::storage().pallet().entry_root(); // iterable
```
We now have:
```rs
runtime::storage().pallet().entry(key1, key2, key3); // only fetchable
runtime::storage().pallet().entry_iter(); // only iterable
runtime::storage().pallet().entry_iter1(key1); // only iterable
runtime::storage().pallet().entry_iter2(key1, key2); // only iterable
```

tadeohepperle committed:
implement partial key iters

tadeohepperle committed:
Merge branch 'master' into tadeo-hepperle-support-iterating-over-n-maps-with-partial-keys

tadeohepperle committed:
format

tadeohepperle committed:
make tests compile

tadeohepperle committed:
fix docs and try example

tadeohepperle committed:
Merge branch 'master' into tadeo-hepperle-support-iterating-over-n-maps-with-partial-keys

lexnv committed:
codegen: Fetch and decode metadata version then fallback (#1092)

* codegen: Fetch and decode metadata version then fallback

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* cli: Add `unstable-metadata` attribute to the subxt macro

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* docs: Add missing comma

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* macro: Add `GenerateRuntimeApi`

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* Update subxt/src/lib.rs

Co-authored-by: James Wilson <james@jsdw.me>

* Update macro/src/lib.rs

Co-authored-by: James Wilson <james@jsdw.me>

* subxt: Adjust docs

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* cli: Import `GenerateRuntimeApi`

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

---------

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>
Co-authored-by: James Wilson <james@jsdw.me>

dependabot[bot] committed:
Bump darling from 0.20.1 to 0.20.3 (#1085)

Bumps [darling](https://github.com/TedDriggs/darling) from 0.20.1 to 0.20.3.
- [Release notes](https://github.com/TedDriggs/darling/releases)
- [Changelog](https://github.com/TedDriggs/darling/blob/master/CHANGELOG.md)
- [Commits](https://github.com/TedDriggs/darling/compare/v0.20.1...v0.20.3)

---
updated-dependencies:
- dependency-name: darling
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: James Wilson <james@jsdw.me>

dependabot[bot] committed:
Bump either from 1.8.1 to 1.9.0 (#1084)

Bumps [either](https://github.com/bluss/either) from 1.8.1 to 1.9.0.
- [Commits](https://github.com/bluss/either/compare/1.8.1...1.9.0)

---
updated-dependencies:
- dependency-name: either
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: James Wilson <james@jsdw.me>

dependabot[bot] committed:
Bump clap from 4.3.11 to 4.3.19 (#1083)

Bumps [clap](https://github.com/clap-rs/clap) from 4.3.11 to 4.3.19.
- [Release notes](https://github.com/clap-rs/clap/releases)
- [Changelog](https://github.com/clap-rs/clap/blob/master/CHANGELOG.md)
- [Commits](https://github.com/clap-rs/clap/compare/v4.3.11...v4.3.19)

---
updated-dependencies:
- dependency-name: clap
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: James Wilson <james@jsdw.me>
Co-authored-by: Tadeo Hepperle <62739623+tadeohepperle@users.noreply.github.com>

dependabot[bot] committed:
Bump trybuild from 1.0.81 to 1.0.82 (#1082)

Bumps [trybuild](https://github.com/dtolnay/trybuild) from 1.0.81 to 1.0.82.
- [Release notes](https://github.com/dtolnay/trybuild/releases)
- [Commits](https://github.com/dtolnay/trybuild/compare/1.0.81...1.0.82)

---
updated-dependencies:
- dependency-name: trybuild
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: James Wilson <james@jsdw.me>

jsdw committed:
Prep for 0.30.1 release (#1094)

ascjones committed:
Set minimum supported `rust-version` to `1.70` (#1097)

dependabot[bot] committed:
Bump serde_json from 1.0.103 to 1.0.104 (#1100)

Bumps [serde_json](https://github.com/serde-rs/json) from 1.0.103 to 1.0.104.
- [Release notes](https://github.com/serde-rs/json/releases)
- [Commits](https://github.com/serde-rs/json/compare/v1.0.103...v1.0.104)

---
updated-dependencies:
- dependency-name: serde_json
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump serde from 1.0.175 to 1.0.179 (#1101)

Bumps [serde](https://github.com/serde-rs/serde) from 1.0.175 to 1.0.179.
- [Release notes](https://github.com/serde-rs/serde/releases)
- [Commits](https://github.com/serde-rs/serde/compare/v1.0.175...v1.0.179)

---
updated-dependencies:
- dependency-name: serde
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jsdw committed:
Tests: support 'substrate-node' too and allow multiple binary paths (#1102)

* Support 'substrate-node' too and allow multiple binary paths

* fmt

* clippy

* fix path

jsdw commented on subxt/examples/storage_iterating_partial.rs:
Just a reminder that any new examples should be linked from the book somewhere (this ensures that they are "valuable" enough to warrant having in the first place, but also means they will be discovered when people read the relevant sections).

Probably the book should be tweaked anyway; it says:

```rust
//! As well as accessing specific entries, some storage locations can also be iterated over (such as
//! the map of account information). To do this, suffix `_root` onto the query constructor 
```

And  possibly at that point we should mention that you can iterate with `_iter`, and for maps with multiple keys, we can `_iter2`, `_iter3` etc to iterate with any number of the map keys. Then, probably we can just have an inline code example showing this (if there are any good examples of storage entries) or just leave the user to discover it if not.

tadeohepperle replied:
I could not find a good example to show, so I removed that file, but I added some explanations to the book, how to use the `_iterX` constructors.

tadeohepperle committed:
adjust book

tadeohepperle committed:
remove the partial iteration example. there was nothing good to show

tadeohepperle committed:
Merge branch 'master' into tadeo-hepperle-support-iterating-over-n-maps-with-partial-keys

tadeohepperle committed:
Merge branch 'master' into tadeo-hepperle-support-iterating-over-n-maps-with-partial-keys

jsdw commented on codegen/src/api/storage.rs:
```diff
@@ -61,147 +62,105 @@ fn generate_storage_entry_fns(
+                    .iter()
+                    .enumerate()
+                    .map(|(i, f)| {
+                        let ident: Ident = format_ident!("_{}", syn::Index::from(i));
```
Nit; is `Ident` type explicitly needed?

tadeohepperle replied:
Ah, yes it is needed because of the `quote!( #field_name: impl ::std::borrow::Borrow<#field_ty> )` in Line 136, where `field_name` is the `ident` from above.

jsdw commented on subxt/src/book/usage/storage.rs:
````diff
@@ -39,19 +39,26 @@
+//! let storage_query = subxt::dynamic::storage_iter("System", "Account");
 //! ```
 //!
+//! Some storage entries require multiple keys to be accessed, for example
+//! `polkadot::storage().preimage().preimage_for(_, _)`, which takes a `H256` and a `u32` as keys.
+//! For these multi-key storage entries, the generated static interface also offers _partial iteration_.
+//! This means iterating over a subset of storage entries, where just a subset of the keys match.
+//! These generated constructors have a suffix `_iter1` / `_iter2` / ... added, to iterate over all
+//! storage entries given the first 1 / 2 / ... keys.
````
Perhaps an example will make this point clearer, eg:

```suggestion
//! Some storage entries are maps with multiple keys. As an example, we might end up with
//! an API like `runtime::storage().foo().bar(u8, bool, u16, String)` to fetch some entry "bar".
//! When this is the case, the codegen will generate multiple iterator query functions alongside
//! the function to fetch an individual value:
//!
//! - `runtime::storage().foo().bar(u8, bool, u16, String)`: fetch a single entry from the "bar" map.
//! - `runtime::storage().foo().bar_iter()`: iterate over all of the entries in the "bar" map.
//! - `runtime::storage().foo().bar_iter1(u8)`: iterate over all of the entries in the "bar" map under 
//!   a given `u8`.
//! - `runtime::storage().foo().bar_iter2(u8, bool)`: iterate over all of the entries in the "bar" map under 
//!   a given `u8` and `bool` value.
//! - `runtime::storage().foo().bar_iter2(u8, bool, u16)`: iterate over all of the entries in the "bar" map under 
//!   a given `u8`, `bool` and `u16` value.
```

jsdw commented on subxt/src/storage/storage_address.rs:
```diff
@@ -229,7 +229,7 @@ pub fn make_static_storage_map_key<T: codec::Encode>(t: T) -> StaticStorageMapKe
 
 /// Construct a new dynamic storage lookup to the root of some entry.
-pub fn dynamic_root(
+pub fn dynamic_iter(
```
Ah, but now the `dynamic_iter` doesn't actually provide the ability to iterate over subsets of keys like the static one does.

Maybe we should just remove this function entirely, since people can just use `dynamic` below with an empty vec to achieve the same, or with a subset of keys to hopefully be able to iterate whatever they like.

jsdw replied:
(and if we remove this, remember to update the guide :)

jsdw commented:
Looks good overall! Just a few small comments ( a load of spaces in CHANGELOG to revert, comment update, remove `dynamic_root` now I think is best).

Also, I had al ook and found these storage entries with multiple values that might be easier to "set up" for testing:

```
multisig.multisigs(AccountId32, [u8; 32])
staking.erasStakers(u32, AccountId32)
xcmPallet.remoteLockedFungibles(u32, AccountId32, XcmVersionedAssetId)
```

Multisigs can be created and so maybe that's not too hard to write a test for? I mainly included the bottom one as the only example I could see of more than 2 keys :)

It would be great to have a test if possible to check that this stuff works (and with the multisig one we could also test the `dynamic` version to check it can do the same.

tadeohepperle committed:
Merge branch 'master' into tadeo-hepperle-support-iterating-over-n-maps-with-partial-keys

tadeohepperle committed:
revert spaces in changelog

tadeohepperle committed:
Support more types in Storage entry constructors  (#1105)

* implement test for type alias being used

* Bump serde from 1.0.179 to 1.0.183 (#1112)

Bumps [serde](https://github.com/serde-rs/serde) from 1.0.179 to 1.0.183.
- [Release notes](https://github.com/serde-rs/serde/releases)
- [Commits](https://github.com/serde-rs/serde/compare/v1.0.179...v1.0.183)

---
updated-dependencies:
- dependency-name: serde
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Bump regex from 1.9.1 to 1.9.3 (#1110)

Bumps [regex](https://github.com/rust-lang/regex) from 1.9.1 to 1.9.3.
- [Release notes](https://github.com/rust-lang/regex/releases)
- [Changelog](https://github.com/rust-lang/regex/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rust-lang/regex/compare/1.9.1...1.9.3)

---
updated-dependencies:
- dependency-name: regex
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* revert yaml changes

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

tadeohepperle committed:
Update subxt/src/book/usage/storage.rs

Co-authored-by: James Wilson <james@jsdw.me>

tadeohepperle committed:
remove dynamic_iter

tadeohepperle committed:
Merge branch 'master' into tadeo-hepperle-support-iterating-over-n-maps-with-partial-keys

tadeohepperle committed:
fix example

tadeohepperle committed:
format

tadeohepperle committed:
add example, adjust book

tadeohepperle committed:
Merge branch 'master' into tadeo-hepperle-support-iterating-over-n-maps-with-partial-keys

lexnv commented on codegen/src/api/storage.rs:
```diff
@@ -61,147 +62,231 @@ fn generate_storage_entry_fns(
+    if let Some(ty) = type_path.vec_type_param() {
+        return quote!([#ty]);
+    }
+    // String is cast to str
```
nice!

lexnv commented on codegen/src/api/storage.rs:
```diff
@@ -61,147 +62,231 @@ fn generate_storage_entry_fns(
-    };
+    let all_fns = (0..=keys.len()).map(|n_keys| {
+        let keys_slice = &keys[..n_keys];
+        let (fn_name, is_fetchable, is_iterable) = if n_keys == keys.len() {
+            let fn_name = format_ident!("{snake_case_name}");
+            (fn_name, true, false)
+        } else {
+            let fn_name = if n_keys == 0 {
+                format_ident!("{snake_case_name}_iter")
+            } else {
+                format_ident!("{snake_case_name}_iter{}", n_keys)
+            };
+            (fn_name, false, true)
+        };
+        let is_fetchable_type = is_fetchable.then_some(quote!(#crate_path::storage::address::Yes)).unwrap_or(quote!(()));
+        let is_iterable_type = is_iterable.then_some(quote!(#crate_path::storage::address::Yes)).unwrap_or(quote!(()));
```
nit: I found this a bit hard to follow, maybe a match would state the intent a bit better.
And since the `is_fetchable` and `is_iterable` are mutually exclusive, we could probably do with just one of them, up to you

jsdw replied:
I had a go and came up with something like this:

```rust
let is_iterator = n_keys != keys.len();

let fn_name = match (is_iterator, keys.len()) {
    (false, _) => format_ident!("{snake_case_name}"),
    (true, 0) => format_ident!("{snake_case_name}_iter"),
    (true, n) => format_ident!("{snake_case_name}_iter{}", n)
};

let yes_ty = || quote!(#crate_path::storage::address::Yes);
let no_ty = || quote!(());

let is_fetchable_type = is_iterator.then(no_ty).unwrap_or_else(yes_ty);
let is_iterable_type = is_iterator.then(yes_ty).unwrap_or_else(no_ty);
```

jsdw reviewed (approved):
Looks great to me :)

lexnv commented on subxt/src/book/usage/storage.rs:
```diff
@@ -39,19 +39,33 @@
+//!   a given `u8`.
+//! - `runtime::storage().foo().bar_iter2(u8, bool)`: iterate over all of the entries in the "bar" map under
+//!   a given `u8` and `bool` value.
+//! - `runtime::storage().foo().bar_iter2(u8, bool, u16)`: iterate over all of the entries in the "bar" map under
```
```suggestion
//! - `runtime::storage().foo().bar_iter3(u8, bool, u16)`: iterate over all of the entries in the "bar" map under
```

tadeohepperle committed:
Update subxt/src/book/usage/storage.rs

Co-authored-by: Alexandru Vasile <60601340+lexnv@users.noreply.github.com>

lexnv commented on subxt/examples/storage_iterating_partial.rs:
I wonder if this partial example should (also) be part of the integration-tests

tadeohepperle replied:
I tried that but had some issues. Let me see again, it is a good idea.

lexnv reviewed (approved):
👍
